src/types_.py

```diff
@@ -3,17 +3,332 @@
     """
 
-    url: str
-    content: str
+    url: Url
+    content: Content
+
+
+NavlinkTitle = str
+
+
+class IndexFile(typing.NamedTuple):
+    """Information about a documentation page.
+
+    Attrs:
+        title: The title for the index.
+        content: The local content of the index.
+    """
+
+    title: NavlinkTitle
+    content: Content | None
+
+
+class Index(typing.NamedTuple):
+    """Information about the local and server index page.
+
+    Attrs:
+        server: The index page on the server.
+        local: The local index file contents.
+    """
+
+    server: Page | None
+    local: IndexFile
+
+
+Level = int
+TablePath = str
+
+
+class PathInfo(typing.NamedTuple):
+    """Represents a file or directory in the docs directory.
+
+    Attrs:
+        local_path: The path to the file on the local disk.
+        level: The number of parent directories to the docs folder including the docs folder.
+        table_path: The computed table path based on the disk path relative to the docs folder.
+        navlink_title: The title of the navlink.
+    """
+
+    local_path: Path
+    level: Level
+    table_path: TablePath
+    navlink_title: NavlinkTitle
+
+
+PathInfoLookup = dict[tuple[Level, TablePath], PathInfo]
+
+
+class Navlink(typing.NamedTuple):
+    """Represents navlink of a table row of the navigation table.
+
+    Attrs:
+        title: The title of the documentation page.
+        link: The relative URL to the documentation page or None if there is no link.
+    """
+
+    title: NavlinkTitle
+    link: str | None
+
+
+class TableRow(typing.NamedTuple):
+    """Represents one parsed row of the navigation table.
+
+    Attrs:
+        level: The number of parents, is 1 if there is no parent.
+        path: The a unique string identifying the row.
+        navlink: The title and relative URL to the documentation page.
+    """
+
+    level: Level
+    path: TablePath
+    navlink: Navlink
+
+    @property
+    def is_group(self) -> bool:
+        """Whether the row is a group of pages."""
+        return self.navlink.link is None
+
+    def to_line(self) -> str:
+        """Convert to a line in the navigation table.
+
+        Returns:
+            The line in the navigation table.
+        """
+        return (
+            f"| {self.level} | {self.path} | [{self.navlink.title}]({self.navlink.link or ''}) |"
+        )
+
+
+TableRowLookup = dict[tuple[Level, TablePath], TableRow]
+
+
+class Action(str, Enum):
+    """The possible actions to take for a page.
+
+    Attrs:
+        CREATE: create a new page.
+        NOOP: no action required.
+        UPDATE: change aspects of an existing page.
+        DELETE: remove an existing page.
+    """
+
+    CREATE = "create"
+    NOOP = "noop"
+    UPDATE = "update"
+    DELETE = "delete"
+
+
+@dataclasses.dataclass
+class BaseAction:
+    """Represents an action on a page.
+
+    Attrs:
+        action: The action to execute on the page.
+    """
+
+    action: Action
```

Comment:
Yes I'm not happy with it either, good suggestion, I'll change it. I'll also change it to `type_` rather than `action_type` because `action.type_` is a little more succinct than `action.action_type`. What do you think?

Reply:
This has now been pushed to this PR

Reply:
Yeap, also had "type" in mind, so perfect 👍